### PR TITLE
feat: bump jQuery to v3.6.0

### PR DESCRIPTION
**Pull request by trivikr**
# Description

Building on top of https://github.com/lsegal/yard/pull/1358

# Testing
The generate API reference can be viewed at https://lsegal-yard-1397.netlify.app/
* The jQuery version is `"3.6.0"`, verified by running `$.fn.jquery` in Console.
* The jQuery Migrate version is `"3.3.2"`, verified by running `$.migrateVersion` in Console.
* All toggle functionalities are working:
  * <details>
    <summary>Table of Contents on landing page</summary>

    https://user-images.githubusercontent.com/16024985/131574652-39808355-b4e5-43ba-918d-37459158e1de.mov

    </details>

  * <details>
    <summary>Inheritance tree on Array page.</summary>

    https://user-images.githubusercontent.com/16024985/131574907-bf3d318e-fa61-4201-95ea-020db86875a3.mov

    </details>
  * <details>
    <summary>Source above footer on Array page.</summary>

    https://user-images.githubusercontent.com/16024985/131574943-d3a06add-7ab9-4c24-be15-091e5b51aa59.mov

    </details>
  * <details>
    <summary>Defines on Module: YARD::Handlers::Ruby page.</summary>

    https://user-images.githubusercontent.com/16024985/131583803-f222164c-f1f5-452e-94b9-00f6114f798d.mov

    </details>

# Completed Tasks

- [x] I have read the [Contributing Guide][contrib].
- [x] The pull request is complete (implemented / written).
- [x] Git commits have been cleaned up (squash WIP / revert commits).
- [x] I wrote tests and ran `bundle exec rake` locally (if code is attached to PR).

[contrib]: https://github.com/lsegal/yard/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by trivikr** [comment]
@lsegal This PR is ready for review.

**Review by AllanZhengYP — APPROVED** [review]

**Comment by mullermp** [comment]
:shipit:

**Comment by trivikr** [comment]
Update: Added jQuery Migrate to not break downstream consumers with custom templates who may use jQuery deprecated APIs.

It's depending on prior feedback in:
* Tweet: https://twitter.com/lsegal/status/1310315871051501568
* PR comment: https://github.com/lsegal/yard/pull/1358#issuecomment-716072243

**Review by alextwoods — APPROVED** [review]
Change looks good to me - Resolves the CVE and I believe this addresses issues with previous jquery upgrades.

File: templates/default/fulldoc/html/js/app.js
Comment (lsegal) [inline templates/default/fulldoc/html/js/app.js:10]: Is there a reason code is being replaced if this is theoretically not a breaking API change?

It would seem more safe to test this without any deltas to app.js to ensure that jquery-migrate is in fact 100% compatible.

Comment (trivikr) [reply templates/default/fulldoc/html/js/app.js:10]: > Is there a reason code is being replaced if this is theoretically not a breaking API change?

The functionality described in toggle breaks, if it's not replaced by click.
You can view ToC which animates hide by default at https://lsegal-yard-jquery-bump-no-edits.netlify.app/

<details>
<summary>Screen recording</summary>

https://user-images.githubusercontent.com/16024985/131713399-a4c58c17-ffcb-49ab-a5e9-5b2e8067a866.mov

</details>

I verified that replacing toggle with click does not break existing functionality even in jQuery 1.7.1
The doc can be viewed at https://lsegal-yard-replace-toggle-with-click.netlify.app/
  * <details>
    <summary>Table of Contents on landing page</summary>

    https://user-images.githubusercontent.com/16024985/131574652-39808355-b4e5-43ba-918d-37459158e1de.mov

    </details>

  * <details>
    <summary>Inheritance tree on Array page.</summary>

    https://user-images.githubusercontent.com/16024985/131574907-bf3d318e-fa61-4201-95ea-020db86875a3.mov

    </details>
  * <details>
    <summary>Source above footer on Array page.</summary>

    https://user-images.githubusercontent.com/16024985/131574943-d3a06add-7ab9-4c24-be15-091e5b51aa59.mov

    </details>
  * <details>
    <summary>Defines on Module: YARD::Handlers::Ruby page.</summary>

    https://user-images.githubusercontent.com/16024985/131583803-f222164c-f1f5-452e-94b9-00f6114f798d.mov

    </details>

Comment (lsegal) [reply templates/default/fulldoc/html/js/app.js:10]: > The functionality described in toggle breaks, if it's not replaced by click. 

If behavior breaks with `toggle()` then this is considered a backwards incompatible change and can't be pulled into YARD without a major version bump unfortunately. 

Downstream consumers are relying on customized templates that may contain use of toggle that is expected to continue to work.

If this were to be accepted, I believe you would need to backport `toggle()` and any major API changes from jquery. I thought that's what jquery.migrate does?

Comment (trivikr) [reply templates/default/fulldoc/html/js/app.js:10]: > I thought that's what jquery.migrate does?

I thought the same, but may be the toggle wrapper is incorrect in jquery-migrate.

I'll post a PR with just commit [`56b30fc` (#1397)](https://github.com/lsegal/yard/pull/1397/commits/56b30fc49cbe3031a62a2e962efc3f263d437387) which replaces toggle with click. That way:
* The functionality introduced by toggle works after switching to click in jQuery 1.7.1
* The downstream users of YARD won't be impacted, as jQuery version is not bumped.
* The users who want to upgrade jQuery can replace `jquery.js` vendored by YARD.

Comment (trivikr) [reply templates/default/fulldoc/html/js/app.js:10]: > I'll post a PR with just commit [`56b30fc` (#1397)](https://github.com/lsegal/yard/pull/1397/commits/56b30fc49cbe3031a62a2e962efc3f263d437387) which replaces toggle with click.

The PR to replace toggle(function, function) with click is posted at https://github.com/lsegal/yard/pull/1399

**Comment by trivikr** [comment]
I'll close this PR after rebase once https://github.com/lsegal/yard/pull/1399 is merged.
This way, we'll have PR to refer to if jQuery upgrade is prioritized in future.

**Comment by lsegal** [comment]
> I'll close this PR after rebase once #1399 is merged.
> This way, we'll have PR to refer to if jQuery upgrade is prioritized in future.

I'm not sure I understand this.

The intention is to backport toggle(), not remove it. #1399 is also not mergeable as is since it too is a breaking change.

The only way any of this can be merged is if the underlying jquery library is API compatible with the previous version distributed by YARD. 

Custom templates (packaged as separate gems) depend on specific functions provided by YARD's templating system in the browser environment. Any changes to those functions would be equivalent to changes in Ruby methods. Replacing YARD's own internal API calls is secondary to this requirement.

TL;DR: toggle() must continue to be available to YARD's JavaScript templating system. Whether that's a custom implementation or a vendored jQuery backport doesn't matter, but the function has to exist.

**Comment by trivikr** [comment]
> The intention is to backport toggle(), not remove it. #1399 is also not mergeable as is since it too is a breaking change.

#1399 is not a breaking change, as it does not touch jQuery version.
It just changes `.toggle(function, function)` calls to equivalent `.click()` calls.
Am I understanding it incorrectly?

> TL;DR: toggle() must continue to be available to YARD's JavaScript templating system.

With #1399, `.toggle(function, function)` will continue to be available to YARD's JavaScript templating system. The PR is just replacing calls to `.toggle(function, function)` from base templates, and not the implementation of the API.

**Comment by lsegal** [comment]
> #1399 is not a breaking change, as it does not touch jQuery version vendored by YARD.
> It just changes `.toggle(function, function)` calls to equivalent `.click()` calls in base templates.
> Am I understanding it incorrectly?

It's not a breaking change per se, but it's also not a change that resolves any specific issue, and doesn't need to be merged. 

> The PR is just replacing API calls to .toggle(function, function) in base templates, and not the implementation of the API.

Sure, but I have the same comment for that issue as I did here: https://github.com/lsegal/yard/pull/1397#discussion_r700391726 -- specifically, toggle() should nto be changed so we can validate that the underlying jQuery API has not changed. Changing the code to use click() would generate a false positive success case where your specific output [test case](https://lsegal-yard-1397.netlify.app/) looks valid but breaks for downstream consumers that still rely on toggle.

Leaving #1399 aside for the moment, this PR is still a breaking change and needs to be resolved before merging. #1399 should not be necessary to resolve the underlying jQuery upgrade and probably won't be merged because of the above comments.

**Comment by lsegal** [comment]
It is also worth noting here that the missing `toggle()` call in jquery.migrate.js indicates to me that it is explicitly not 100% API compatible with the current distributed version.

Do we know what other APIs have been changed/removed? A full list of changes would be useful to understand. Presumably every one of those functions would also need to be backported, otherwise we risk breaking downstream template authors that may depend on other functions outside of the obvious `toggle()` example.

**Comment by trivikr** [comment]
> It's not a breaking change per se, but it's also not a change that resolves any specific issue, and doesn't need to be merged.

The #1399 allows users of YARD to bump jquery to v3.6.0, as the new implementation `.click()` is forward compatible. It's explained in the Background and Description section.

> Leaving #1399 aside for the moment, this PR is still a breaking change and needs to be resolved before merging.

I think there's a confusion. I don't plan to merge this PR.
I will just close it, as we would be unblocked once #1399 is merged.

**Pull request closed without merging**
